Locality and Geocoorddetail batch import tool

melton-jason opened the pull request:
Fixes #4319

The Locality and GeoCoordDetail batch import tool allows users to match and update the Latitude1, Longitidude1, and Datum fields for Locality records and associate new GeoCoordDetails for the matched Locality records. 

The primary motivation for the tool is such that the georefrencing for existing records in the database can take place outside of Specify (such as with [CoGe](https://coge.geo-locate.org/About.htm)), and there is no easy way to currently repatriate the data back into Specify and update those records. 

This process is facilitated by requiring the guids of the Locality records in the CSV. Specify will use these guids to match/find the corresponding Locality record(s). 
With the Locality record, the tool does the following: 
- If any `latitude1`, `longitude1`, and/or `datum` fields have values for the Locality record in the CSV, Specify will overwrite the current values in the database with those specified in the CSV
- If any other column in the CSV matches a non-relationship field name in the GeoCoordDetail table, Specify will delete the previous GeoCoordDetail associated with the Locality record and create a new one populated with field values from the CSV
- If there are any other columns in the CSV, Specify will display a warning to the user indicating there are unknown columns, but allow the user to proceed; the unknown columns will be ignored if the user proceeds. 

---
### Demo
Currently, access to the tool can be found in the User Tools menu: 

<img width="1436" alt="locality_import_button" src="https://github.com/specify/specify7/assets/64045831/20c8556e-35c3-4b71-a6e7-2f82ba4b5de2">


The primary interface is very similar to that of importing a file to be used as a DataSet in the WorkBench. Specify supports different encodings and delimiters, and displays the first 100 rows of the file:

<img width="1427" alt="coge_preview" src="https://github.com/specify/specify7/assets/64045831/a3181a88-26be-4a1c-bf07-33461e7b9607">

The `Import File` button can be pressed to initiate the parsing/upload. 

If there are any unknown columns in the dataset, a warning will first be displayed to the user so they can decide how to proceed: 

<img width="1427" alt="coge_column_warning" src="https://github.com/specify/specify7/assets/64045831/6d6e4b1a-9b5e-43d0-b5f8-f6e910cfc54a">

If there are any bad values discovered when parsing the results, Specify will display a downloadable message stating which line the error was found on along with a message with more information about how to resolve the parsing error:

<img width="1436" alt="parse_error" src="https://github.com/specify/specify7/assets/64045831/93576706-98ce-4b26-ac2c-31d4511fa0a0">


If the parsing and upload is successful, Specify will display a brief results page stating how many records of each table were affected, along with the option to create a RecordSet of modified Locality Records:

<img width="1436" alt="upload_results" src="https://github.com/specify/specify7/assets/64045831/3e813a55-28fb-44d2-ae53-ce40ed310ef3">

--- 

### Checklist

- [x] Self-review the PR after opening it to make sure the changes look good
      and self-explanatory (or properly documented)
- [ ] Add automated tests
- [x] Add relevant issue to release milestone

### Testing instructions

For testing purposes, here is an example set of data for use on the `KUFish` database: 
[kufish_coge_sample.csv](https://github.com/specify/specify7/files/15257843/kufish_coge_sample.csv)

Other datasets have been made for `FITZ_NHMD`, `fwri`, `KUBirds`, and `lsumzmammals` and are available in the following folder in the google drive: 
https://drive.google.com/drive/folders/1B0hKQMBaX82nBUOB95uEa3H0MIrGrBv7?usp=drive_link

A foundation for other Datasets for use in other databases can be made by creating a Locality -> GeoCoordDetail query and exporting the results. For an example, here is the Query used to populate most of `kufish_coge_sample.csv`: 
https://kufish51623-edge.test.specifysystems.org/specify/query/191/

There should be no errors in the `kufish_coge_sample` dataset presently, so modifications can be made to accomplish the desired errors and edge cases. 

Generally, the cases which would need to be tested are: 
- [ ] Ensure an error message is displayed to the user when there is no Locality which has a guid which exists in the dataset
- [ ]  Data in unknown columns is properly ignored and not accounted for when uploading
- [ ] If a Locality contains existing data for `GeoCoordDetails`, it should be deleted if there is at least one `GeoCoordDetail` field in the dataset
- [ ] If there are no values in any GeoCoordDetail field for a Locality in the dataset and the Locality contains a GeoCoordDetail in Specify, the GeoCoordDetail should not be overwritten
- [ ] Ensure that only the Locality fields which contain data in the dataset get updated for a Locality in the dataset (i.e., if `longitude1` and `datum` have values in the dataset but `latitude1` is either not a column or empty, ensure that `latitude1` is never overwritten) 
- [ ] Ensure parsing error messages are intuitive enough to diagnose and resolve the problem with the dataset
  - Potential ways to introduce invalid values include: 
    - Entering a value in a field which exceeds the maximum length for the field (the maximum length for a field can be found in the Schema Configuration) 
    - Entering an incorrect type of value for the field (such as entering a letter into a number/integer field)
    - Not following a UIFormatter for a field

maxpatiiuk commented on specifyweb/frontend/js_src/lib/localization/header.ts:
```diff
@@ -143,6 +143,9 @@ export const headerText = createDictionary({
     `,
   },
+  coGeImportDataset: {
+    'en-us': 'Import CoGe Dataset'
```
oh, so CoGe is even exposed to the user?
would they know what coge means?

maxpatiiuk commented on specifyweb/frontend/js_src/lib/localization/locality.ts:
```diff
@@ -292,4 +292,17 @@ export const localityText = createDictionary({
     'de-ch': 'DD MM SS.SS N/S/O/W (32 45 42.84 N)',
   },
+  localityImportHeaderError: {
+    'en-us': 'Errors Found in Column Headers',
```
(optional) is this better or worse?
```suggestion
    'en-us': 'Errors found in column headers',
```

maxpatiiuk commented on specifyweb/frontend/js_src/lib/components/Header/userToolDefinitions.ts:
```diff
@@ -99,6 +99,14 @@ const rawUserTools = ensure<IR<IR<Omit<MenuItem, 'name'>>>>()({
     },
   },
+  [commonText.import()]: {
+    coGeImport: {
```
aah, I just realized that CoGe is a name of an org
feel free to disregard my comments about it then

melton-jason commented on specifyweb/frontend/js_src/lib/components/Header/ImportLocalitySet.tsx:
```diff
@@ -0,0 +1,133 @@
+              <Button.DialogClose>{commonText.close()}</Button.DialogClose>
+              {headerErrors.missingRequiredHeaders.length === 0 && (
+                <Submit.Save
+                  onClick={() =>
```
Fixed in https://github.com/specify/specify7/pull/4548/commits/09c58810c2b8a5bb2fd1905dd694437fa8deb08e

melton-jason commented on specifyweb/frontend/js_src/lib/components/Header/ImportLocalitySet.tsx:
```diff
@@ -0,0 +1,133 @@
+              </>
+            )}
+            <H2>{localityText.localityImportedAcceptedHeaders()}</H2>
+            <p>{Array.from(acceptedHeaders).join(', ')}</p>
```
Fixed in https://github.com/specify/specify7/pull/4548/commits/09c58810c2b8a5bb2fd1905dd694437fa8deb08e

melton-jason commented on specifyweb/frontend/js_src/lib/components/Header/ImportLocalitySet.tsx:
```diff
@@ -0,0 +1,133 @@
+            {headerErrors.unrecognizedHeaders.length > 0 && (
+              <>
+                <H2>{localityText.localityImportUnrecognizedHeaders()}</H2>
+                <p>{headerErrors.unrecognizedHeaders.join(', ')}</p>
```
Fixed in https://github.com/specify/specify7/pull/4548/commits/09c58810c2b8a5bb2fd1905dd694437fa8deb08e

melton-jason commented on specifyweb/frontend/js_src/lib/components/Header/ImportLocalitySet.tsx:
```diff
@@ -0,0 +1,133 @@
+            {headerErrors.missingRequiredHeaders.length > 0 && (
+              <>
+                <H2>{localityText.localityImportMissingHeader()}</H2>
+                <p>{headerErrors.missingRequiredHeaders.join(', ')}</p>
```
Fixed in https://github.com/specify/specify7/pull/4548/commits/09c58810c2b8a5bb2fd1905dd694437fa8deb08e

melton-jason commented on specifyweb/frontend/js_src/lib/components/Header/userToolDefinitions.ts:
```diff
@@ -99,6 +99,14 @@ const rawUserTools = ensure<IR<IR<Omit<MenuItem, 'name'>>>>()({
+  [commonText.import()]: {
+    coGeImport: {
+      title: headerText.coGeImportDataset(),
+      enabled: () => true,
```
Fixed in https://github.com/specify/specify7/pull/4548/commits/09c58810c2b8a5bb2fd1905dd694437fa8deb08e

melton-jason commented on specifyweb/frontend/js_src/lib/components/Molecules/FilePicker.tsx:
```diff
@@ -174,3 +192,319 @@ export const fileToText = async (
     fileReader.readAsText(file, encoding);
   });
+
+export function CsvFilePicker({
```
Fixed in https://github.com/specify/specify7/pull/4548/commits/09c58810c2b8a5bb2fd1905dd694437fa8deb08e

melton-jason commented on specifyweb/frontend/js_src/lib/components/Molecules/FilePicker.tsx:
```diff
@@ -174,3 +192,319 @@ export const fileToText = async (
+      <div className="w-96">
+        <FilePicker
+          acceptedFormats={['.csv', '.tsv', '.psv', '.txt']}
+          onFileSelected={(file): void => setFile(file)}
```
Fixed in https://github.com/specify/specify7/pull/4548/commits/09c58810c2b8a5bb2fd1905dd694437fa8deb08e

melton-jason commented on specifyweb/frontend/js_src/lib/components/Molecules/FilePicker.tsx:
```diff
@@ -174,3 +192,319 @@ export const fileToText = async (
+      preview={preview}
+      onFileImport={(hasHeader): void => {
+        if (!Array.isArray(preview)) {
+          console.error('Failed to parse data for File ', file.name, preview);
```
Fixed in https://github.com/specify/specify7/pull/4548/commits/09c58810c2b8a5bb2fd1905dd694437fa8deb08e

melton-jason commented on specifyweb/frontend/js_src/lib/components/Molecules/FilePicker.tsx:
```diff
@@ -174,3 +192,319 @@ export const fileToText = async (
+        onChange={({ data }): void => handleChange(data)}
+        onCleared={(): void => {
+          setState(undefined);
+        }}
```
Fixed in https://github.com/specify/specify7/pull/4548/commits/09c58810c2b8a5bb2fd1905dd694437fa8deb08e

maxpatiiuk commented on specifyweb/frontend/js_src/lib/components/QueryBuilder/CreateRecordSet.tsx:
```diff
@@ -38,14 +37,19 @@ export function CreateRecordSet({
 
+  const resolvedRecordIds = React.useMemo(
+    () => (typeof recordIds === 'function' ? recordIds() : recordIds),
+    [recordIds]
```
the getIds function provided as a prop to CreateRecordSetFromQuery is re-created on each render, so this useMemo would be called on each render
is that an issue?

maxpatiiuk commented on specifyweb/frontend/js_src/lib/components/QueryBuilder/CreateRecordSet.tsx:
```diff
@@ -109,3 +115,30 @@ export function CreateRecordSet({
+  readonly queryResource: SpecifyResource<SpQuery> | undefined;
+}): JSX.Element {
+  const recordSetName =
+    queryResource === undefined || queryResource.isNew()
+      ? undefined
+      : queryResource.get('name');
```
(optional)

why does CreateRecordSetFromQuery exist?
is it just for these 3 lines?
the rest is just proxying the props to CreateRecordSet

can you just call CreateRecordSet directly - that would take less code that this

the only reason for CreateRecordSetFromQuery to exist is if you want to add some documentation for it, that is not applicable for CreateRecordSet

melton-jason replied:
> can you just call CreateRecordSet directly - that would take less code that this

Fixed in https://github.com/specify/specify7/pull/4548/commits/b844b88dd40bfd6544237b010650775e653bebff

maxpatiiuk commented on specifyweb/frontend/js_src/lib/components/WorkBench/resultsParser.ts:
```diff
@@ -226,6 +215,32 @@ export function resolveValidationMessage(
+      value: payload.value as string,
+      formatter: payload.formatter as string,
+    });
+  else return undefined;
```
(optional) return `key` instead + a console.error?

maxpatiiuk commented on specifyweb/frontend/js_src/lib/localization/locality.ts:
```diff
@@ -292,4 +292,42 @@ export const localityText = createDictionary({
+    'en-us': 'Only the following headers are accepted',
+  },
+  localityImportErrorDialogHeader: {
+    'en-us': 'Error(s) Occured while Parsing Dataset',
```
we normally don't use Header Case in dialog headers
maybe we should, but for concistency we should either do so everywhere, or nowhere

maxpatiiuk commented on specifyweb/frontend/js_src/lib/localization/locality.ts:
```diff
@@ -292,4 +292,42 @@ export const localityText = createDictionary({
+  },
+  localityImportErrorDialogDetails: {
+    'en-us':
+      'The following Errors occured at the provided Row (Line) Numbers of the file while parsing the Dataset: ',
```
remove trailing space

maxpatiiuk replied:
also, I don't see usages of this string, though maybe github UI just didn't load that file

maxpatiiuk commented on specifyweb/frontend/js_src/lib/utils/ajax/definitions.ts:
```diff
@@ -84,4 +85,6 @@ export const httpCodeToErrorMessage: RR<ValueOf<typeof Http>, string> = {
   [Http.INSUFFICIENT_STORAGE]: `
     This error likely happened because the server has run out of storage space.
   `,
+  [Http.UNPROCESSABLE]:
```
could you insert httpCodeToErrorMessage in the same order as in Http?

melton-jason replied:
Fixed in f0e372fb81dd0a622c8c7410995e677237b0097f

maxpatiiuk commented on specifyweb/specify/views.py:
```diff
@@ -768,3 +768,145 @@ def abort_merge_task(request, merge_id: int) -> http.HttpResponse:
+        },
+        "responses": {
+            "200": {
+                "description": "The Locality records were updated and GeocoordDetails uploaded successfully ",
```
trailing space

maxpatiiuk commented on specifyweb/specify/views.py:
```diff
@@ -768,3 +768,145 @@ def abort_merge_task(request, merge_id: int) -> http.HttpResponse:
+                                            "message": {
+                                                "description": "Keys for errors which occured during parsing",
+                                                "type": "string",
+                                                "enum": localityParseErrorMessages
```
smart 👍

maxpatiiuk commented on specifyweb/specify/views.py:
```diff
@@ -768,3 +768,145 @@ def abort_merge_task(request, merge_id: int) -> http.HttpResponse:
+                    "schema": {
+                        "type": "object",
+                        "properties": {
+                            "columnHeaders": {
```
(optional) add more documentation or example of what column headers are expected

alesan99 reviewed (changes requested):
Looks like Geo Coord Details are overwritten / created when there are no Geo Coord values in the row in the csv.

![chrome_VAvikgqy6w](https://github.com/specify/specify7/assets/8108710/49ff9bcf-c506-49d2-add5-9b61bac9bf4b)
TO
![chrome_xN4ZhSVrI5](https://github.com/specify/specify7/assets/8108710/fbbda2fd-3abf-4f4f-96c8-4c4177080b9b)
Tested on `KUBirds_8_25_23` with `kubirds_coge_sample.csv ` (after erasing the Geo Coord Detail values from a row)

All the other tests did work good though

melton-jason commented:
> Looks like Geo Coord Details are overwritten / created when there are no Geo Coord values in the row in the csv.
> 
> ![chrome_VAvikgqy6w](https://private-user-images.githubusercontent.com/8108710/330144004-49ff9bcf-c506-49d2-add5-9b61bac9bf4b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU4NzA2MTEsIm5iZiI6MTcxNTg3MDMxMSwicGF0aCI6Ii84MTA4NzEwLzMzMDE0NDAwNC00OWZmOWJjZi1jNTA2LTQ5ZDItYWRkNS05YjYxYmFjOWJmNGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MTZUMTQzODMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Njg5ZDAwN2M2MjlhY2U5MjlhODY5M2E1NzYwMWUyMzJmZGJjZmUzMWU0MzRhNzBjMjc4YzcwYmQzYTY5NjQ5NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.lZeefdwyO538UxFpMM0ORG-0lnCwls6Gc_x4korCaWE) TO ![chrome_xN4ZhSVrI5](https://private-user-images.githubusercontent.com/8108710/330141982-fbbda2fd-3abf-4f4f-96c8-4c4177080b9b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU4NzA2MTEsIm5iZiI6MTcxNTg3MDMxMSwicGF0aCI6Ii84MTA4NzEwLzMzMDE0MTk4Mi1mYmJkYTJmZC0zYWJmLTRmNGYtOTZjOC00YzQxNzcwODBiOWIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MTZUMTQzODMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTFlYzI3MjhlZjU4NTAyZmE2MWMwYjBmMDNiNzRhZmE3ZjZkZmI0ZGExOTYxODZmZWNlYzEzZjJkNzFjODdmMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Weakn_sG6kRW3819cjSEfX9zp03QuWLNEqMyhHBoKwc) Tested on `KUBirds_8_25_23` with `kubirds_coge_sample.csv ` (after erasing the Geo Coord Detail values from a row)
> 
> All the other tests did work good though

https://github.com/specify/specify7/pull/4548#pullrequestreview-2053252737

@alesan99 
I couldn't seem to recreate this issue. 

To test this, I used a single Locality record on KUFish with a GUID of `72fc97f4-2926-4f35-aad3-0306b382483c` and inserted some values in the GeoCoordDetail. 

Then I imported the following CSV using the tool (where the only values are in the `GUID, latitude1, longitude1, and datum` columns): 
```csv
GUID,Latitude1,Longitude1,Datum,Source,georefdetdate,georefremarks,maxuncertaintyest,maxuncertaintyestunit,namedplaceextent,errorpolygon,uncertaintypolygon
72fc97f4-2926-4f35-aad3-0306b382483c,17.0432700000,-96.6793200000,WGS84,,,,,,,,
```

The value in the GeoCoordDetail for the Locality remained untouched. 

Are you able to provide the line of the CSV which corresponds to a Locality/GeoCoordDetails for which this occurs?

emenslin reviewed (changes requested):
### Testing instructions

For testing purposes, here is an example set of data for use on the `KUFish` database: 
[kufish_coge_sample.csv](https://github.com/specify/specify7/files/15257843/kufish_coge_sample.csv)

Other datasets have been made for `FITZ_NHMD`, `fwri`, `KUBirds`, and `lsumzmammals` and are available in the following folder in the google drive: 
https://drive.google.com/drive/folders/1B0hKQMBaX82nBUOB95uEa3H0MIrGrBv7?usp=drive_link

A foundation for other Datasets for use in other databases can be made by creating a Locality -> GeoCoordDetail query and exporting the results. For an example, here is the Query used to populate most of `kufish_coge_sample.csv`: 
https://kufish51623-edge.test.specifysystems.org/specify/query/191/

There should be no errors in the `kufish_coge_sample` dataset presently, so modifications can be made to accomplish the desired errors and edge cases. 

Generally, the cases which would need to be tested are: 
- [x] Ensure an error message is displayed to the user when there is no Locality which has a guid which exists in the dataset
- [x]  Data in unknown columns is properly ignored and not accounted for when uploading
- [x] If a Locality contains existing data for `GeoCoordDetails`, it should be deleted if there is at least one `GeoCoordDetail` field in the dataset
- [x] If there are no values in any GeoCoordDetail field for a Locality in the dataset and the Locality contains a GeoCoordDetail in Specify, the GeoCoordDetail should not be overwritten
- [x] Ensure that only the Locality fields which contain data in the dataset get updated for a Locality in the dataset (i.e., if `longitude1` and `datum` have values in the dataset but `latitude1` is either not a column or empty, ensure that `latitude1` is never overwritten) 
- [ ] Ensure parsing error messages are intuitive enough to diagnose and resolve the problem with the dataset
  - Potential ways to introduce invalid values include: 
    - Entering a value in a field which exceeds the maximum length for the field (the maximum length for a field can be found in the Schema Configuration) 
    - Entering an incorrect type of value for the field (such as entering a letter into a number/integer field)
    - Not following a UIFormatter for a field

## Issues 
The original error message that happens if there is an unknown field stays behind the loading screen the whole time and is still behind the create record set dialog at the end as well. This dialog should close once you start the upload.

https://github.com/specify/specify7/assets/150940840/60fb437c-117b-49af-bcd3-fa135f00b95f

Could not consistently recreate this but sometimes after you imported a file and got some importing error (e.g field value is too long) if you tried to import the file again you would get a full specify error where you would have to return to the homepage. 

## Changes 
I think when you are importing instead of just a loading screen for a long time it should have something like the workbench has 
Like this: 
![Screenshot 2024-05-16 131607](https://github.com/specify/specify7/assets/150940840/1713ddab-9316-45e6-8296-3b8e2fe79908)

I also think that it should auto create a record set. With the current behavior it is very easy to accidentally click out and then you cannot go back to create one again. At the very least I think you should not be able to click out of the 'create record set' dialog unless you explicitly click 'cancel'.

I also think the error dialog needs to include the header- just seeing the row isn't always super intuitve, also the column names in this dialog are too close together in my opinion. 
![Screenshot 2024-05-16 130541](https://github.com/specify/specify7/assets/150940840/3550e734-c7b1-4da5-8410-4955f570b9bb)

alesan99 commented:
> @alesan99 I couldn't seem to recreate this issue.
> 
> To test this, I used a single Locality record on KUFish with a GUID of `72fc97f4-2926-4f35-aad3-0306b382483c` and inserted some values in the GeoCoordDetail.
> 
> Then I imported the following CSV using the tool (where the only values are in the `GUID, latitude1, longitude1, and datum` columns):
> 
> ```
> GUID,Latitude1,Longitude1,Datum,Source,georefdetdate,georefremarks,maxuncertaintyest,maxuncertaintyestunit,namedplaceextent,errorpolygon,uncertaintypolygon
> 72fc97f4-2926-4f35-aad3-0306b382483c,17.0432700000,-96.6793200000,WGS84,,,,,,,,
> ```
> 
> The value in the GeoCoordDetail for the Locality remained untouched.
> 
> Are you able to provide the line of the CSV which corresponds to a Locality/GeoCoordDetails for which this occurs?

Hm after doing some more testing in KU_Birds it looks like it only happens if I had a source defined in the `source` column, having only the columns you specified indeed does not cause any errors.

This was the culprit:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/a114s239/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/a114s239/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"Aptos Narrow", sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
-->
</style>
</head>

<body link="#467886" vlink="#96607D">


guid | latitude1 | longitude1 | datum | source | georefdetdate | georefremarks | maxuncertaintyest | maxuncertaintyestunit | namedplaceextent | errorpolygon | uncertaintypolygon
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
5d8f344a-7b48-11e4-8ef3-782bcb9cd5b5 |   |   |  WGS84 | BioGeomancer |   |   |   |   |   |   |  



</body>

</html>

Removing the value from source no longer overwrites the Geo Coord details.

I don't quite know "source"'s relation to Geo Coord details, but unless this is unintentional it looks like nothing was going wrong after all.

melton-jason commented:
@specify/ux-testing 

Here are some changes I have made since the last review: 
- Removed all specific references of CoGe
  - The tool is source-agnostic and only requires the data match its own specific format. Thus all terms of 'CoGe Dataset' I have renamed to 'Locality Repatriation Dataset'. 
  - Non-usable bonus points if you can come up with a better name!

<img width="1436" alt="locality_import_button" src="https://github.com/specify/specify7/assets/64045831/70aad25a-a5bf-4c7d-b65b-070809f76dda">

- In the Parse Error dialog, I have included the field name as a column

<img width="1436" alt="parse_error" src="https://github.com/specify/specify7/assets/64045831/7f3fe626-ca3a-43af-8546-9fbab17f38c5">

- On a successful import, a recordset is always created, and in the import/upload results a link to the recordset is provided

<img width="1436" alt="upload_results" src="https://github.com/specify/specify7/assets/64045831/d2a26f2f-8fcc-4a51-a557-fbddc7e3f4f6">

- Fixed "The original error message that happens if there is an unknown field stays behind the loading screen the whole time and is still behind the create record set dialog at the end as well. This dialog should close once you start the upload"
  - From  https://github.com/specify/specify7/pull/4548#pullrequestreview-2061752787

---

Can these aspects be tested and feedback provided? 

As far as including a progress bar over a loading is concerned, is there any additional functionality requested? We could integrate with the Notifications system, for example.

Areyes42 reviewed (changes requested):
### Testing instructions

For testing purposes, here is an example set of data for use on the `KUFish` database:  
[kufish\_coge\_sample.csv](https://github.com/specify/specify7/files/15257843/kufish_coge_sample.csv)

Other datasets have been made for `FITZ_NHMD`, `fwri`, `KUBirds`, and `lsumzmammals` and are available in the following folder in the google drive:  
[https://drive.google.com/drive/folders/1B0hKQMBaX82nBUOB95uEa3H0MIrGrBv7?usp=drive\_link](https://drive.google.com/drive/folders/1B0hKQMBaX82nBUOB95uEa3H0MIrGrBv7?usp=drive_link)

A foundation for other Datasets for use in other databases can be made by creating a Locality -> GeoCoordDetail query and exporting the results. For an example, here is the Query used to populate most of `kufish_coge_sample.csv`:  
[https://kufish51623-edge.test.specifysystems.org/specify/query/191/](https://kufish51623-edge.test.specifysystems.org/specify/query/191/)

There should be no errors in the `kufish_coge_sample` dataset presently, so modifications can be made to accomplish the desired errors and edge cases.

Generally, the cases which would need to be tested are:

-  [x]  Ensure an error message is displayed to the user when there is no Locality which has a guid which exists in the dataset
-  [x] Data in unknown columns is properly ignored and not accounted for when uploading
-  [x]  If a Locality contains existing data for `GeoCoordDetails`, it should be deleted if there is at least one `GeoCoordDetail` field in the dataset
-  [x]  If there are no values in any GeoCoordDetail field for a Locality in the dataset and the Locality contains a GeoCoordDetail in Specify, the GeoCoordDetail should not be overwritten
-  [x]  Ensure that only the Locality fields which contain data in the dataset get updated for a Locality in the dataset (i.e., if `longitude1` and `datum` have values in the dataset but `latitude1` is either not a column or empty, ensure that `latitude1` is never overwritten)
-  [ ]  Ensure parsing error messages are intuitive enough to diagnose and resolve the problem with the dataset
    -   Potential ways to introduce invalid values include:
        -   Entering a value in a field which exceeds the maximum length for the field (the maximum length for a field can be found in the Schema Configuration)
        -   Entering an incorrect type of value for the field (such as entering a letter into a number/integer field)
        -   Not following a UIFormatter for a field
***
### Import tool looks excellent!

For the most part, I agree with everything that @emenslin mentioned for changes. I also think 'Import Locality Dataset' is a more concise name instead of 'Import Locality Repatriation Dataset'.

I noticed one small discrepancy: the row number in the parsing error notification does not match the one in the actual dataset. I understand that this happens because the row count [starts at `0`](https://skillcrush.com/blog/why-programmers-start-counting-at-zero/) and doesn't account if the first row is selected as the header. Although it's a nitpicky suggestion, I believe matching the row numbers would make it more intuitive and easier to navigate errors, since most spreadsheet editors start with 1.
![Screenshot 2024-05-28 at 12 19 12 PM](https://github.com/specify/specify7/assets/130881183/000da30a-0c9d-49a9-95aa-1fde975ebca7)

melton-jason commented:
@specify/ux-testing 
I will update the testing instructions/documentation with the changes soon, but this is ready to be tested again! 

In short: 
- Now instead of the loading dialog while the process takes place, a dialog will be displayed which shows the state of the process

- If the process is Parsing or Uploading the Data Set, the operation can be cancelled

- Notifications will be sent to the user which initiated the process about the state of the process, and can view the Results or Errors at any time as long as the Notification is not deleted. 
  - If the recordset associated with a success is deleted or otherwise not found, it is replaced in the notification with a `Browse In Forms` button which acts like the Query Result button with the same name (save for there is also a button in the header of the Browse In Forms which allows a new record set to be created) 
  
- The `rowNumber` attribute in the Failure/Error dialog and export now start their indexing at 1 rather than 0

emenslin reviewed (changes requested):
### Testing instructions

For testing purposes, here is an example set of data for use on the `KUFish` database: 
[kufish_coge_sample.csv](https://github.com/specify/specify7/files/15257843/kufish_coge_sample.csv)

Other datasets have been made for `FITZ_NHMD`, `fwri`, `KUBirds`, and `lsumzmammals` and are available in the following folder in the google drive: 
https://drive.google.com/drive/folders/1B0hKQMBaX82nBUOB95uEa3H0MIrGrBv7?usp=drive_link

A foundation for other Datasets for use in other databases can be made by creating a Locality -> GeoCoordDetail query and exporting the results. For an example, here is the Query used to populate most of `kufish_coge_sample.csv`: 
https://kufish51623-edge.test.specifysystems.org/specify/query/191/

There should be no errors in the `kufish_coge_sample` dataset presently, so modifications can be made to accomplish the desired errors and edge cases. 

Generally, the cases which would need to be tested are: 
- [x] Ensure an error message is displayed to the user when there is no Locality which has a guid which exists in the dataset
- [x]  Data in unknown columns is properly ignored and not accounted for when uploading
- [x] If a Locality contains existing data for `GeoCoordDetails`, it should be deleted if there is at least one `GeoCoordDetail` field in the dataset
- [x] If there are no values in any GeoCoordDetail field for a Locality in the dataset and the Locality contains a GeoCoordDetail in Specify, the GeoCoordDetail should not be overwritten
- [x] Ensure that only the Locality fields which contain data in the dataset get updated for a Locality in the dataset (i.e., if `longitude1` and `datum` have values in the dataset but `latitude1` is either not a column or empty, ensure that `latitude1` is never overwritten) 
- [x] Ensure parsing error messages are intuitive enough to diagnose and resolve the problem with the dataset
  - Potential ways to introduce invalid values include: 
    - Entering a value in a field which exceeds the maximum length for the field (the maximum length for a field can be found in the Schema Configuration) 
    - Entering an incorrect type of value for the field (such as entering a letter into a number/integer field)
    - Not following a UIFormatter for a field

Functionality still seems to be working great and I love the new loading dialog and the notifications system!

### Changes 

- If the data set is parsing you cannot actually cancel it. The cancel button is there but when you press it it doesn't do anything


https://github.com/specify/specify7/assets/150940840/497770f6-af0d-446b-9f9a-1ea289129003


- The error dialog could still use some work. I am glad that the numbers now line up but I still think it could look a bit better. I think I mentioned this before but changing 'row number' to just 'row' could help and expanding what the minimum width of the dialog can be could also help. Everything just feels too close together 

![Screenshot 2024-06-05 140649](https://github.com/specify/specify7/assets/150940840/a612a522-dbd5-4b46-b254-06e8302037af)

- Along with that, as I mentioned in one of my previous reviews (I think) if there is an error with the import I don't think you should be able to click on import again. Although now you can cancel it I still think the button should just be grayed out until you import a new data set 

https://github.com/specify/specify7/assets/150940840/24e752d1-ce9b-4577-bd54-aea29bc5cbd1

- Sometimes the import gets stuck on 0 and the bar is full and it just sits there for several seconds without changing to the next dialog. This isn't necessarily a big deal but I wanted to bring it up in case it ended up being something important (clicking on notifications does not fix it I just happened to click on it in this video to see if it showed up in the notifications) 

https://github.com/specify/specify7/assets/150940840/a2a9a370-c821-4b17-8e4e-a7c25c39bf8f

- Lastly when the test panel crashed the data set never uploaded. When I tried to recreate it by causing the upload to stop in other ways it always seemed to finish the upload anyway and I got the notification. Not sure if this matters since it was just from the test panel crashing but I wanted to bring it up in case there's more underlying issues 

This is starting to look a lot better though, I'm happy with how far it's come already!

emenslin commented:
I didn't include this in my original review but can you also merge the parsing and the starting loading dialogs into just one progress bar?

emenslin commented:
Results are incorrect these are the results from the current branch (impossible result, there's only 518 rows and there can't be more geocoorddetails than localities) 

https://github.com/specify/specify7/assets/150940840/0926808f-1262-424f-a696-8e10922167a1

This is from the same data set from a previous review 
![Screenshot 2024-05-28 085232](https://github.com/specify/specify7/assets/150940840/209aafe2-8452-4c66-bfd7-a06e75685184)

emenslin reviewed (changes requested):
### Testing instructions

For testing purposes, here is an example set of data for use on the `KUFish` database: 
[kufish_coge_sample.csv](https://github.com/specify/specify7/files/15257843/kufish_coge_sample.csv)

Other datasets have been made for `FITZ_NHMD`, `fwri`, `KUBirds`, and `lsumzmammals` and are available in the following folder in the google drive: 
https://drive.google.com/drive/folders/1B0hKQMBaX82nBUOB95uEa3H0MIrGrBv7?usp=drive_link

A foundation for other Datasets for use in other databases can be made by creating a Locality -> GeoCoordDetail query and exporting the results. For an example, here is the Query used to populate most of `kufish_coge_sample.csv`: 
https://kufish51623-edge.test.specifysystems.org/specify/query/191/

There should be no errors in the `kufish_coge_sample` dataset presently, so modifications can be made to accomplish the desired errors and edge cases. 

Generally, the cases which would need to be tested are: 
- [x] Ensure an error message is displayed to the user when there is no Locality which has a guid which exists in the dataset
- [x]  Data in unknown columns is properly ignored and not accounted for when uploading
- [x] If a Locality contains existing data for `GeoCoordDetails`, it should be deleted if there is at least one `GeoCoordDetail` field in the dataset
- [x] If there are no values in any GeoCoordDetail field for a Locality in the dataset and the Locality contains a GeoCoordDetail in Specify, the GeoCoordDetail should not be overwritten
- [x] Ensure that only the Locality fields which contain data in the dataset get updated for a Locality in the dataset (i.e., if `longitude1` and `datum` have values in the dataset but `latitude1` is either not a column or empty, ensure that `latitude1` is never overwritten) 
- [x] Ensure parsing error messages are intuitive enough to diagnose and resolve the problem with the dataset
  - Potential ways to introduce invalid values include: 
    - Entering a value in a field which exceeds the maximum length for the field (the maximum length for a field can be found in the Schema Configuration) 
    - Entering an incorrect type of value for the field (such as entering a letter into a number/integer field)
    - Not following a UIFormatter for a field

It's looking great so far and fuctionality is still working!

--- 
I did not really notice any big issues so I think we are finally nearing the end so yippee for that. A few minor things I noticed 

- One I know you're working on but I want to include it here for future reference, I think the import button should be disabled after you've tried importing the data set and recieved an error 
- Not sure if this is possible but is there anyway to get rid of the loading dialog when you cancel the parsing? Not a big deal but it just flashing really quick which is not ideal in my opinion and the dialog continues for a minute behind the loading screen before it cancels. Not sure if it's even possible to fix but I figured I'd include it 


https://github.com/specify/specify7/assets/150940840/f823b383-e78b-426f-a98c-3203e7eb5c03


- Also another small thing but the 'first row is header' checkbox does not work in this tool as the first row has to be the header. Not a big deal if it stays in as I imagine users just wouldn't touch it but if it isn't too hard I think it should be removed 


https://github.com/specify/specify7/assets/150940840/96369693-93aa-4a4e-81ca-17b975035ddf


- Last small thing but could we make the headers of the error dialog bold and/or bigger just so they stand out more, I feel like they just look like the rest of the rows 
![Screenshot 2024-06-06 144034](https://github.com/specify/specify7/assets/150940840/caa67023-48fd-4c45-8af8-bd047f81b2f8)


--- 
Also I know I said I wanted the parsing and importing dialog to be merged into one but after testing I think it would be better to go back to 2 dialogs. The parsing takes up the first 50% and the importing the second 50% this means that if there are any errors in the data set the dialog will always hit 50% and then throw the error. Because of this I think that the 2 dialogs makes more sense so that it will just throw the error after the first dialog hits 100% 

https://github.com/specify/specify7/assets/150940840/a8f6c2d9-8aa7-444b-989e-c67939e45bfd


I am open to other opinions so if anyone prefers having just one dialog then we can discuss what the best solution would be or just keep it if more people like the single dialog

realVinayak commented:
@melton-jason I'll highly recommend fixing the failing unit test before UX testing (don't know why that hasn't been looked since, even your first commit failed). At least make an attempt to solve it. If it is an "immediately obvious fix", then do it straight away. Why make the same mistake as v7.9?

melton-jason commented:
@specify/ux-testing 

Now, instead of directly parsing and then uploading the file, a dialog should first appear with the total number of Locality records matched and Geocoorddetail records which will be destroyed/created. 

<img width="870" alt="Screenshot 2024-06-13 at 8 39 36 AM" src="https://github.com/specify/specify7/assets/64045831/7bc5ffe3-c3cc-448a-af5a-93075f5388e6">

At this dialog, the user can "Close" (which will cancel the process), or "Import" (goes ahead with the import process). 

Documentation will be updated shortly!

emenslin reviewed (changes requested):
Broken :( 

https://github.com/specify/specify7/assets/150940840/d65a1bcb-3250-48eb-b305-71cdb061d1b1

realVinayak commented on specifyweb/specify/import_locality.py:
```diff
@@ -0,0 +1,488 @@
+                    geocoorddetails += 1
+            LocalityImportRowResult.objects.create(
+                localityimport=li,
+                rownumber=-1,
```
Simplify some things here. Why `-1`? that's hacky and just bad. What if we want to add a feature that allows users to parse the dataset, and then open the matched localities in a new tab? Then, we need to have the ids of the localities. your current code blocks it. what's worse is that if we add this in the future, then we need to handle the ugly case that there is _some_ dataset out there with `-1`.

Suggested refactor. Do this incrementally.

**Get rid of the state "ParseError" and "Uploaded".** 

Yes, I mean even from upload call. Instead, _infer_ it later on. A very good reason for this is what if there is a business rule or an exception that happens at the DB level? Your code doesn't have any capacity to handle that. And, it is not even that hard to support it. + this means there is no concept of ParseError and Uploaded, so the entire second branch of this function isn't needed. Instead, just make results a `List[Union[UploadedResult, ParsedResult]`. Here, a row can either be parsed or uploaded. Actually, it is almost similar to what you have with `ParsedRow`. Just make add another key `errors`. Change your `parse_locality_set` to do so. That is, parse_locality_set will return `List[ParsedResult]`

```
class ParsedBase(NamedTuple):
    row_number: Optional[int]
    locality: Dict[str, Any]
    geocoorddetail: Optional[Dict[str, Any]]

class ParsedSuccess(ParsedBase):
    locality_id: int

class ParsedFail(ParseBase):
    errors: List[ParseError] 

ParsedRow = ParsedFail | ParsedSuccess

```

Right before upload, check if any row is ParsedFail. If so, skip the upload. Otherwise, on upload success, it will be `UploadSuccessRow`.

So, the entire results will be `List[Union[ParsedFail, Union[ParsedSuccess, UploadSuccessRow]]]`

The benefit of this approach is that the entire `isinstance(results, tuple)` can be removed (can be combined with the later branch). In the future, if we want to include business rules, then this will be trivial (convert failure to ParsedFail). 

Will also simplify 
```
        if results['type'] == 'ParseError':
            status = LocalityImportStatus.FAILED
            for error in results['errors']:
                result = error.to_json()
                LocalityImportRowResult.objects.create(
                    localityimport=li,
                    rownumber=error.row_number,
                    result=json.dumps(result, cls=DjangoJSONEncoder)
                )

        elif results['type'] == 'Uploaded':
            status = LocalityImportStatus.SUCCEEDED
            localities = []
            for index, row in enumerate(results['results']):
                row_number = index + 1
                localities.append(row['locality'])

                LocalityImportRowResult.objects.create(
                    localityimport=li,
                    rownumber=row_number,
                    result=json.dumps(row, cls=DjangoJSONEncoder)
                )
```
to be
```
 for index, row in enumerate(results):
            LocalityImportRowResult.objects.create(
                    localityimport=li,
                    rownumber=row_number,
                    result=json.dumps(row, cls=DjangoJSONEncoder)
                )
            if isinstance(row, UploadSuccessRow):
               localities.append(row['locality'])
```

melton-jason replied:
The rownumber issue has been addressed, but the refactoring of `parse_locality_set` may need to wait as this PR is holding up the release. 
I will open an issue for this.

emenslin reviewed (approved):
### Testing instructions

For testing purposes, here is an example set of data for use on the `KUFish` database: 
[kufish_coge_sample.csv](https://github.com/specify/specify7/files/15257843/kufish_coge_sample.csv)

Other datasets have been made for `FITZ_NHMD`, `fwri`, `KUBirds`, and `lsumzmammals` and are available in the following folder in the google drive: 
https://drive.google.com/drive/folders/1B0hKQMBaX82nBUOB95uEa3H0MIrGrBv7?usp=drive_link

A foundation for other Datasets for use in other databases can be made by creating a Locality -> GeoCoordDetail query and exporting the results. For an example, here is the Query used to populate most of `kufish_coge_sample.csv`: 
https://kufish51623-edge.test.specifysystems.org/specify/query/191/

There should be no errors in the `kufish_coge_sample` dataset presently, so modifications can be made to accomplish the desired errors and edge cases. 

Generally, the cases which would need to be tested are: 
- [x] Ensure an error message is displayed to the user when there is no Locality which has a guid which exists in the dataset
- [x]  Data in unknown columns is properly ignored and not accounted for when uploading
- [x] If a Locality contains existing data for `GeoCoordDetails`, it should be deleted if there is at least one `GeoCoordDetail` field in the dataset
- [x] If there are no values in any GeoCoordDetail field for a Locality in the dataset and the Locality contains a GeoCoordDetail in Specify, the GeoCoordDetail should not be overwritten
- [x] Ensure that only the Locality fields which contain data in the dataset get updated for a Locality in the dataset (i.e., if `longitude1` and `datum` have values in the dataset but `latitude1` is either not a column or empty, ensure that `latitude1` is never overwritten) 
- [x] Ensure parsing error messages are intuitive enough to diagnose and resolve the problem with the dataset
  - Potential ways to introduce invalid values include: 
    - Entering a value in a field which exceeds the maximum length for the field (the maximum length for a field can be found in the Schema Configuration) 
    - Entering an incorrect type of value for the field (such as entering a letter into a number/integer field)
    - Not following a UIFormatter for a field

Fuctionality is working as expected, looks great! 🎉

Areyes42 reviewed (approved):
### Testing instructions

For testing purposes, here is an example set of data for use on the `KUFish` database:  
[kufish\_coge\_sample.csv](https://github.com/specify/specify7/files/15257843/kufish_coge_sample.csv)

Other datasets have been made for `FITZ_NHMD`, `fwri`, `KUBirds`, and `lsumzmammals` and are available in the following folder in the google drive:  
[https://drive.google.com/drive/folders/1B0hKQMBaX82nBUOB95uEa3H0MIrGrBv7?usp=drive\_link](https://drive.google.com/drive/folders/1B0hKQMBaX82nBUOB95uEa3H0MIrGrBv7?usp=drive_link)

A foundation for other Datasets for use in other databases can be made by creating a Locality -> GeoCoordDetail query and exporting the results. For an example, here is the Query used to populate most of `kufish_coge_sample.csv`:  
[https://kufish51623-edge.test.specifysystems.org/specify/query/191/](https://kufish51623-edge.test.specifysystems.org/specify/query/191/)

There should be no errors in the `kufish_coge_sample` dataset presently, so modifications can be made to accomplish the desired errors and edge cases.

Generally, the cases which would need to be tested are:

-  [x]  Ensure an error message is displayed to the user when there is no Locality which has a guid which exists in the dataset
-   [x] Data in unknown columns is properly ignored and not accounted for when uploading
-   [x] If a Locality contains existing data for `GeoCoordDetails`, it should be deleted if there is at least one `GeoCoordDetail` field in the dataset
-   [x] If there are no values in any GeoCoordDetail field for a Locality in the dataset and the Locality contains a GeoCoordDetail in Specify, the GeoCoordDetail should not be overwritten
-   [x] Ensure that only the Locality fields which contain data in the dataset get updated for a Locality in the dataset (i.e., if `longitude1` and `datum` have values in the dataset but `latitude1` is either not a column or empty, ensure that `latitude1` is never overwritten)
-   [x] Ensure parsing error messages are intuitive enough to diagnose and resolve the problem with the dataset
    -  Potential ways to introduce invalid values include:
        -   Entering a value in a field which exceeds the maximum length for the field (the maximum length for a field can be found in the Schema Configuration)
        -   Entering an incorrect type of value for the field (such as entering a letter into a number/integer field)
        -  Not following a UIFormatter for a field
***
I tested with several datasets, and they all look good. Overall functionality looks great, awesome job! 👍

alesan99 reviewed (approved):
Generally, the cases which would need to be tested are: 
- [x] Ensure an error message is displayed to the user when there is no Locality which has a guid which exists in the dataset
- [x]  Data in unknown columns is properly ignored and not accounted for when uploading
- [x] If a Locality contains existing data for `GeoCoordDetails`, it should be deleted if there is at least one `GeoCoordDetail` field in the dataset
- [x] If there are no values in any GeoCoordDetail field for a Locality in the dataset and the Locality contains a GeoCoordDetail in Specify, the GeoCoordDetail should not be overwritten
- [x] Ensure that only the Locality fields which contain data in the dataset get updated for a Locality in the dataset (i.e., if `longitude1` and `datum` have values in the dataset but `latitude1` is either not a column or empty, ensure that `latitude1` is never overwritten) 
- [x] Ensure parsing error messages are intuitive enough to diagnose and resolve the problem with the dataset
  - Potential ways to introduce invalid values include: 
    - Entering a value in a field which exceeds the maximum length for the field (the maximum length for a field can be found in the Schema Configuration) 
    - Entering an incorrect type of value for the field (such as entering a letter into a number/integer field)
    - Not following a UIFormatter for a field

My last review https://github.com/specify/specify7/pull/4548#issuecomment-2120693279 was intended behavior and I haven't encountered anything like data overwriting since.

Errors are nicely displayed and export correctly, and all importing functionality is working.
Mostly tested with KUBirds_8_25_23.

Looks good! 👍👍

combs-a reviewed (changes requested):
![image](https://github.com/specify/specify7/assets/71564869/29b3ded7-32d4-4c4d-9b62-86a09b4e7f53)

Was doing some light testing on this and got this error. These coordinates should be fine in Specify, since Locality is fine with it. Looks like the issue is just the >= in the `parse_latlong` function in `parse.py` when it should just be >!

Good work so far btw Jason! While I haven't done full testing yet, what I've gotten through so far is great. :+1:

melton-jason commented:
Hi @specify/ux-testing!
I have merged production into this PR. Notably, this merge included changes from https://github.com/specify/specify7/pull/4764, which included a lot of changes. 

Also, if you encountered the Issue from https://github.com/specify/specify7/pull/4998 in this PR before, it should now additionally be fixed. 

---

> Was doing some light testing on this and got this error. These coordinates should be fine in Specify, since Locality is fine with it. Looks like the issue is just the >= in the parse_latlong function in parse.py when it should just be >!

From https://github.com/specify/specify7/pull/4548#pullrequestreview-2123922418

Both the WorkBench and the Locality Update Tool utilizes the same parsing/validating code. (So this Issue is really the same as #4914). 
This has been fixed as well!

lexiclevenger reviewed (changes requested):
Generally, the cases which would need to be tested are: 
- [x] Ensure an error message is displayed to the user when there is no Locality which has a guid which exists in the dataset
- [x]  Data in unknown columns is properly ignored and not accounted for when uploading
- [x] If a Locality contains existing data for `GeoCoordDetails`, it should be deleted if there is at least one `GeoCoordDetail` field in the dataset
- [ ] If there are no values in any GeoCoordDetail field for a Locality in the dataset and the Locality contains a GeoCoordDetail in Specify, the GeoCoordDetail should not be overwritten
- [x] Ensure that only the Locality fields which contain data in the dataset get updated for a Locality in the dataset (i.e., if `longitude1` and `datum` have values in the dataset but `latitude1` is either not a column or empty, ensure that `latitude1` is never overwritten) 
- [x] Ensure parsing error messages are intuitive enough to diagnose and resolve the problem with the dataset
  - Potential ways to introduce invalid values include: 
    - Entering a value in a field which exceeds the maximum length for the field (the maximum length for a field can be found in the Schema Configuration) 
    - Entering an incorrect type of value for the field (such as entering a letter into a number/integer field)
    - Not following a UIFormatter for a field
---
The latitude/longitude issue has been fixed! However, It looks like cells that are empty are overwriting fields that contain data.

Before Import:
<img width="1199" alt="Screenshot 2024-06-18 at 12 32 58 PM" src="https://github.com/specify/specify7/assets/164079735/d5d66e0f-830e-48dd-8437-7b9d1a908416">


Imported (blank) value:
<img width="1648" alt="Screenshot 2024-06-18 at 12 33 19 PM" src="https://github.com/specify/specify7/assets/164079735/fd041d8f-2e54-4c35-8748-67d966f90a8a">


After Import
<img width="1204" alt="Screenshot 2024-06-18 at 12 33 36 PM" src="https://github.com/specify/specify7/assets/164079735/ef02acbd-f03c-497e-ac26-0632f6bfca0d">

[Specify 7 System Information - 2024-06-18T17_40_05.394Z.txt](https://github.com/user-attachments/files/15890441/Specify.7.System.Information.-.2024-06-18T17_40_05.394Z.txt)

lexiclevenger reviewed (approved):
My previous comment was the expected behavior; everything looks good!

melton-jason commented on specifyweb/frontend/js_src/lib/components/Header/userToolDefinitions.ts:
```diff
@@ -99,6 +99,14 @@ const rawUserTools = ensure<IR<IR<Omit<MenuItem, 'name'>>>>()({
+    coGeImport: {
+      title: headerText.coGeImportDataset(),
+      enabled: () => true,
+      url: '/specify/import/from-coge/',
```
All references to CoGe have been removed! 
The tool is source-agnostic and only requires that the input file matches an expected format

Areyes42 reviewed (approved):
Generally, the cases which would need to be tested are:

-  [x]  Ensure an error message is displayed to the user when there is no Locality which has a guid which exists in the dataset
-   [x] Data in unknown columns is properly ignored and not accounted for when uploading
-  [x]  If a Locality contains existing data for `GeoCoordDetails`, it should be deleted if there is at least one `GeoCoordDetail` field in the dataset
-  [x]  If there are no values in any GeoCoordDetail field for a Locality in the dataset and the Locality contains a GeoCoordDetail in Specify, the GeoCoordDetail should not be overwritten
-  [x]  Ensure that only the Locality fields which contain data in the dataset get updated for a Locality in the dataset (i.e., if `longitude1` and `datum` have values in the dataset but `latitude1` is either not a column or empty, ensure that `latitude1` is never overwritten)
-  [x]  Ensure parsing error messages are intuitive enough to diagnose and resolve the problem with the dataset
    -   Potential ways to introduce invalid values include:
        -   Entering a value in a field which exceeds the maximum length for the field (the maximum length for a field can be found in the Schema Configuration)
        -   Entering an incorrect type of value for the field (such as entering a letter into a number/integer field)
        -   Not following a UIFormatter for a field
***
Looks good! I also tested previous issues, and it seems like they all got fixed!

combs-a reviewed (changes requested):
Generally, the cases which would need to be tested are: 
- [X] Ensure an error message is displayed to the user when there is no Locality which has a guid which exists in the dataset
- [X]  Data in unknown columns is properly ignored and not accounted for when uploading
- [X] If a Locality contains existing data for `GeoCoordDetails`, it should be deleted if there is at least one `GeoCoordDetail` field in the dataset
- [X] If there are no values in any GeoCoordDetail field for a Locality in the dataset and the Locality contains a GeoCoordDetail in Specify, the GeoCoordDetail should not be overwritten
- [X] Ensure that only the Locality fields which contain data in the dataset get updated for a Locality in the dataset (i.e., if `longitude1` and `datum` have values in the dataset but `latitude1` is either not a column or empty, ensure that `latitude1` is never overwritten) 
- [ ] Ensure parsing error messages are intuitive enough to diagnose and resolve the problem with the dataset
  - Potential ways to introduce invalid values include: 
    - Entering a value in a field which exceeds the maximum length for the field (the maximum length for a field can be found in the Schema Configuration) 
    - Entering an incorrect type of value for the field (such as entering a letter into a number/integer field)
    - Not following a UIFormatter for a field

Some of the error messages have typos/etc, such as below:
![6_18_batchlocal1](https://github.com/specify/specify7/assets/71564869/4a4edfe7-cd91-43ae-abb7-4f2ef0406951)

Additionally, I couldn't do the Locality import on a user account that is a Specify7 Admin but didn't have a role defined in the collection:

https://github.com/specify/specify7/assets/71564869/86634df7-0b45-42ef-bb1a-2666d9261b79

Still, the majority of functionality is there and looking good Jason. Nice work! 🎉

melton-jason commented:
The typos and "Additionally, I couldn't do the Locality import on a user account that is a Specify7 Admin but didn't have a role defined in the collection" have been fixed! 

Previously, the backend was checking if the user was a specify 6 admin, not specify 7 admin...

combs-a reviewed (approved):
Looks good now! The issue with institution admins w/o an admin role in the collection not being able to upload is fixed now. 🎉

maxpatiiuk commented on specifyweb/frontend/js_src/lib/components/LocalityUpdate/index.tsx:
```diff
@@ -0,0 +1,189 @@
+  }
+
+  return (
+    <ProtectedAction action="%" resource="%">
```
uh interesting
is this to restict this to institutional admins only?
if so, could you put a short comment here?

maxpatiiuk commented on specifyweb/frontend/js_src/lib/components/LocalityUpdate/types.ts:
```diff
@@ -0,0 +1,101 @@
+};
+
+export type LocalityUpdateTaskStatus =
+  | 'ABORTED'
```
for consistency, we usually name states in PascalCase not UPPER_CASE

maxpatiiuk commented on specifyweb/frontend/js_src/lib/components/LocalityUpdate/types.ts:
```diff
@@ -0,0 +1,101 @@
+  readonly rowNumber: number;
+};
+
+export type LocalityUpdateTaskStatus =
```
no need to duplicate it. can infer it

```suggestion
export type LocalityUpdateTaskStatus = LocalityUpdateState['state'];
```

maxpatiiuk commented on specifyweb/frontend/js_src/lib/components/LocalityUpdate/types.ts:
```diff
@@ -0,0 +1,101 @@
+
+export type LocalityUpdateState =
+  | State<
+      'ABORTED',
+      { readonly taskstatus: 'ABORTED'; readonly taskinfo: string }
```
hmm, this duplicates the fields
the `State<>` helper adds a `{state:'ABORTED'}`, and then you redundantly add `taskstatus`
Ideally you would pick one or another
If you want to keep the taskstatus name, then get rid of `State<>`

maxpatiiuk commented on specifyweb/frontend/js_src/lib/components/LocalityUpdate/utils.ts:
```diff
@@ -0,0 +1,69 @@
+  } else if (key === 'multipleLocalitiesWithGuid') {
+    return localityText.multipleLocalitiesWithGuid({
+      guid: payload.guid as string,
+      localityIds: (payload.localityIds as RA<number>).join(', '),
```
call formatConjunction instead of join?

maxpatiiuk commented on specifyweb/frontend/js_src/lib/components/Merging/types.ts:
```diff
@@ -1,6 +1,6 @@
-export type MergeStatus = 'ABORTED' | 'FAILED' | 'MERGING' | 'SUCCEEDED';
+export type MergingStatus = 'ABORTED' | 'FAILED' | 'MERGING' | 'SUCCEEDED';
```
(optional) use PascalCase for statuses?

maxpatiiuk commented on specifyweb/frontend/js_src/lib/components/QueryBuilder/CreateRecordSet.tsx:
```diff
@@ -116,29 +127,16 @@ export function CreateRecordSet({
-    queryResource === undefined || queryResource.isNew()
-      ? undefined
-      : queryResource.get('name');
+function LoadingDialog(): JSX.Element {
```
(optional) add optional `header` parameter to `<LoadingScreen />` since we needed that in several places
and then look for usages of `loadingBar` in dialog that can be replaced with it (there would be one in the react workbench I think)

maxpatiiuk commented on specifyweb/frontend/js_src/lib/components/WorkBench/Results.tsx:
```diff
@@ -80,6 +81,38 @@ export function WbUploaded({
+    ]
+  ) => ValueOf<Partial<Record<Lowercase<keyof Tables>, number>>>;
+}): JSX.Element {
+  const resolvedRecords =
+    typeof rawSortFunction === 'function'
+      ? Object.entries(recordCounts).sort(sortFunction(rawSortFunction))
+      : Object.entries(recordCounts);
```
useMemo?

maxpatiiuk reviewed (commented):
sorry that I didn't get to reviewing this yesterday.
looking great!
